dashboard/app/models/user.rb

```diff
@@ -804,28 +804,40 @@ def update_with_password(params, *options)
   end
 
   def update_primary_contact_info(user: {email: nil, hashed_email: nil})
-    email = user[:email]
-    hashed_email = user[:hashed_email]
-
-    return false if email.nil? && hashed_email.nil?
-    return false if teacher? && email.nil?
+    new_email = user[:email]
+    new_hashed_email = new_email.present? ? User.hash_email(new_email) : user[:hashed_email]
+
+    return false if new_email.nil? && new_hashed_email.nil?
+    return false if teacher? && new_email.nil?
+
+    # If an auth option already exists with this email, it beocmes the primary.
+    # Otherwise make a new one.
+    existing_auth_option = authentication_options.find_by hashed_email: new_hashed_email
```

Comment:
Assumes data consistency, which I'd like to do moving forward under multi-auth.  Any good reason we shouldn't _always_ search only by hashed_email?

Reply:
i'm definitely in favor of this being our goal. my only worry is that `hash_email` isn't called until `before_save` (rather than recalculated every time password is changed)

Reply:
As of https://github.com/code-dot-org/code-dot-org/commit/2a28ce8a9d183f01676898c341669e35d6a51202, in `AuthenticationOption` it runs `before_validation` instead of `before_save`.

https://github.com/code-dot-org/code-dot-org/blob/3e3bdba0b96e563821e3d4e64f1e90b7a2a7d6bc/dashboard/app/models/authentication_option.rb#L28-L30

I think this would be fine either way though - when calling `authentication_options.find_by` we should only be checking against saved rows (so `hashed_email` will be filled) and I made an adjustment at the top of this method to ensure `new_hashed_email` is also present and appropriately hashed.